http/src/handler.rs

```diff
@@ -70,87 +63,227 @@ impl<M: Metadata, S: Middleware<M>> server::Handler<HttpStream> for ServerHandle
+							Some(result) => Response::ok(format!("{}\n", result)),
+						}.into()))
+					},
+					Ok(Async::NotReady) => RpcPollState::NotReady(RpcHandlerState::Waiting(waiting)),
+					Err(_) => RpcPollState::Ready(RpcHandlerState::Writing(Response::internal_error())),
+				}
+			},
+			state => RpcPollState::NotReady(state),
+		};
+
+		let (new_state, is_ready) = new_state.decompose();
+		match new_state {
+			RpcHandlerState::Writing(res) => {
+				let mut response: server::Response = res.into();
+				let cors_header = mem::replace(&mut self.cors_header, cors::CorsHeader::Invalid);
+				Self::set_response_headers(response.headers_mut(), self.is_options, cors_header.into());
+				Ok(Async::Ready(response))
+			},
+			state => {
+				self.state = state;
+				if is_ready {
+					self.poll()
+				} else {
+					Ok(Async::NotReady)
+				}
+			},
+		}
+	}
 }
 
 impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
-	fn response_headers(&self, is_options: bool, cors_header: Option<header::AccessControlAllowOrigin>) -> Headers {
-		let mut headers = Headers::new();
-		headers.set(self.response.content_type.clone());
+	fn read_headers(
+		&self,
+		request: server::Request,
+		continue_on_invalid_cors: bool,
+	) -> RpcHandlerState<M> {
+		if self.cors_header == cors::CorsHeader::Invalid && !continue_on_invalid_cors {
+			return RpcHandlerState::Writing(Response::invalid_cors());
+		}
+		// Read metadata
+		let metadata = self.jsonrpc_handler.extractor.read_metadata(&request);
 
+		// Proceed
+		match *request.method() {
+			// Validate the ContentType header
+			// to prevent Cross-Origin XHRs with text/plain
+			Method::Post if Self::is_json(request.headers().get::<header::ContentType>()) => {
+				RpcHandlerState::ReadingBody {
+					metadata: metadata,
+					request: Default::default(),
+					body: request.body(),
+				}
+			},
+			// Just return error for unsupported content type
+			Method::Post => {
+				RpcHandlerState::Writing(Response::unsupported_content_type())
+			},
+			// Don't validate content type on options
+			Method::Options => {
+				RpcHandlerState::Writing(Response::empty())
+			},
+			// Disallow other methods.
+			_ => {
+				RpcHandlerState::Writing(Response::method_not_allowed())
+			},
+		}
+	}
+
+	fn read_body(
+		&self,
+		mut body: hyper::Body,
+		mut request: Vec<u8>,
+		metadata: M,
+	) -> Result<RpcPollState<M>, hyper::Error> {
+		loop {
+			match body.poll()? {
+				// TODO [ToDr] reject too large requests?
+				Async::Ready(Some(chunk)) => {
+					request.extend_from_slice(&*chunk)
+				},
+				Async::Ready(None) => {
+					let content = match ::std::str::from_utf8(&request) {
+						Ok(content) => content,
+						Err(_) => {
+							// returns empty response on invalid utf8
+							return Ok(RpcPollState::Ready(RpcHandlerState::Writing(Response::empty())));
+						},
+					};
+
+					return Ok(RpcPollState::Ready(RpcHandlerState::Waiting(
+						self.jsonrpc_handler.handler.handle_request(content, metadata)
```

Comment:
this function is called `read_body`. imo it should not handle the request, just read the body ;)